File: src/lib/battle/equipment-system.ts
/**
 * Ship Equipment System
 *
 * Modular equipment that can be installed on ships to modify stats,
 * add special abilities, and provide tactical advantages in combat.
 *
 * Templates and slot configuration are in equipment-templates.ts
 */

import type { DamageTypes, ResistanceTypes, StatusEffectType } from './damage-types'
import type { AdvancedCombatUnit, UnitClass } from './advanced-unit'

// Import and re-export templates and config from dedicated module
import { DEFAULT_SLOT_CONFIG, EQUIPMENT_TEMPLATES, getTemplatesBySlot, getTemplatesByRarity, getTemplatesForClass, getTemplateKeys } from './equipment-templates'
export { DEFAULT_SLOT_CONFIG, EQUIPMENT_TEMPLATES, getTemplatesBySlot, getTemplatesByRarity, getTemplatesForClass, getTemplateKeys }

// ============================================================================
// EQUIPMENT TYPES
// ============================================================================

/**
 * Equipment slot types
 */
export type EquipmentSlot =
  | 'weapon'        // Primary weapon systems
  | 'secondary'     // Secondary/utility weapons
  | 'shield'        // Shield generators
  | 'armor'         // Armor plating
  | 'engine'        // Propulsion systems
  | 'computer'      // Targeting/hacking systems
  | 'special'       // Special equipment
  | 'consumable'    // One-time use items

/**
 * Equipment rarity tiers
 */
export type EquipmentRarity =
  | 'common'        // Basic equipment
  | 'uncommon'      // Improved versions
  | 'rare'          // Advanced technology
  | 'epic'          // Experimental tech
  | 'legendary'     // Unique/artifact equipment

/**
 * Equipment activation type
 */
export type ActivationType =
  | 'passive'       // Always active
  | 'active'        // Must be activated, has cooldown
  | 'triggered'     // Activates on certain conditions
  | 'consumable'    // Single use

// ============================================================================
// EQUIPMENT DEFINITION
// ============================================================================

/**
 * Stat modifiers from equipment
 */
export interface EquipmentStatModifiers {
  // Multiplicative bonuses (1.0 = no change)
  damageMultiplier?: number
  shieldMultiplier?: number
  armorMultiplier?: number
  hullMultiplier?: number
  speedMultiplier?: number

  // Additive bonuses
  accuracyBonus?: number
  evasionBonus?: number
  critChanceBonus?: number
  critDamageBonus?: number
  pointDefenseBonus?: number
  shieldRegenBonus?: number

  // Damage type bonuses
  damageTypeBonuses?: Partial<DamageTypes>

  // Resistance bonuses
  resistanceBonuses?: Partial<ResistanceTypes>

  // Crew bonuses
  crewCapacityBonus?: number
  crewCombatBonus?: number
  boardingPowerBonus?: number
  antiBoardingBonus?: number
}

/**
 * Active ability definition
 */
export interface EquipmentAbility {
  /** Ability identifier */
  id: string
  /** Display name */
  name: string
  /** Description */
  description: string
  /** Cooldown in combat rounds */
  cooldown: number
  /** Current cooldown remaining */
  currentCooldown: number
  /** Energy/resource cost */
  energyCost: number
  /** Duration in rounds (0 = instant) */
  duration: number
  /** Effect of the ability */
  effect: AbilityEffect
}

/**
 * Ability effect types
 */
export type AbilityEffect =
  | DamageAbilityEffect
  | BuffAbilityEffect
  | DebuffAbilityEffect
  | RepairAbilityEffect
  | SpecialAbilityEffect

export interface DamageAbilityEffect {
  type: 'damage'
  damage: Partial<DamageTypes>
  targetCount: number // Number of targets (0 = all enemies)
  areaEffect: boolean
  penetration: number // Ignores this % of resistance
}

export interface BuffAbilityEffect {
  type: 'buff'
  target: 'self' | 'ally' | 'all_allies'
  modifiers: Partial<EquipmentStatModifiers>
  duration: number
}

export interface DebuffAbilityEffect {
  type: 'debuff'
  target: 'enemy' | 'all_enemies'
  statusEffect: StatusEffectType
  duration: number
  strength: number
}

export interface RepairAbilityEffect {
  type: 'repair'
  target: 'self' | 'ally' | 'all_allies'
  repairAmount: number // Percentage of max
  repairType: 'shield' | 'armor' | 'hull' | 'all'
}

export interface SpecialAbilityEffect {
  type: 'special'
  effectId: string
  params: Record<string, unknown>
}

/**
 * Triggered effect condition
 */
export interface TriggerCondition {
  /** Trigger type */
  trigger:
    | 'on_hit'           // When this unit hits enemy
    | 'on_damage'        // When this unit takes damage
    | 'on_kill'          // When this unit destroys enemy
    | 'on_critical'      // When scoring critical hit
    | 'on_shield_break'  // When shields are depleted
    | 'on_low_hull'      // When hull drops below threshold
    | 'on_round_start'   // At the start of each round
    | 'on_round_end'     // At the end of each round
    | 'on_boarding'      // When initiating/defending boarding
  /** Threshold for trigger (e.g., 30 for "below 30% hull") */
  threshold?: number
  /** Chance to trigger (0-100) */
  chance: number
}

/**
 * Complete equipment definition
 */
export interface Equipment {
  /** Unique identifier */
  id: string
  /** Display name */
  name: string
  /** Description */
  description: string
  /** Equipment slot */
  slot: EquipmentSlot
  /** Rarity tier */
  rarity: EquipmentRarity
  /** Activation type */
  activationType: ActivationType

  /** Compatible unit classes */
  compatibleClasses: UnitClass[]

  /** Stat modifiers (always active for passive) */
  statModifiers: EquipmentStatModifiers

  /** Active abilities */
  abilities: EquipmentAbility[]

  /** Triggered effects */
  triggers: Array<{
    condition: TriggerCondition
    effect: AbilityEffect
  }>

  /** Installation requirements */
  requirements: EquipmentRequirements

  /** Cost to install */
  installCost: {
    metal: number
    crystal: number
    deuterium: number
  }

  /** Power consumption */
  powerConsumption: number

  /** Weight/mass */
  mass: number

  /** Number of uses remaining (for consumables) */
  usesRemaining?: number
}

/**
 * Installation requirements
 */
export interface EquipmentRequirements {
  /** Minimum tech levels */
  techLevels?: {
    weaponsTech?: number
    shieldTech?: number
    armorTech?: number
    computerTech?: number
    engineTech?: number
  }
  /** Minimum ship level/experience */
  minShipLevel?: number
  /** Required other equipment */
  requiredEquipment?: string[]
  /** Incompatible equipment */
  incompatibleEquipment?: string[]
}

// ============================================================================
// EQUIPPED UNIT
// ============================================================================

/**
 * Equipment installed on a unit
 */
export interface InstalledEquipment {
  equipment: Equipment
  slot: EquipmentSlot
  slotIndex: number
  active: boolean
}

/**
 * Unit with equipment
 */
export interface EquippedUnit {
  unit: AdvancedCombatUnit
  equipment: InstalledEquipment[]
  totalPowerConsumption: number
  maxPower: number
  slotConfig: Record<EquipmentSlot, number>
}

// ============================================================================
// EQUIPMENT FUNCTIONS
// ============================================================================

/**
 * Create equipment instance from template
 */
export function createEquipment(
  templateId: string,
  instanceId?: string
): Equipment | null {
  const template = EQUIPMENT_TEMPLATES[templateId]
  if (!template) return null

  return {
    id: instanceId ?? `${templateId}_${Date.now()}`,
    ...template,
    // Deep copy abilities to avoid shared state
    abilities: template.abilities.map(a => ({ ...a })),
    triggers: template.triggers.map(t => ({
      condition: { ...t.condition },
      effect: { ...t.effect } as AbilityEffect,
    })),
  }
}

/**
 * Check if equipment can be installed on unit
 */
export function canInstallEquipment(
  equipment: Equipment,
  unit: AdvancedCombatUnit,
  currentEquipment: InstalledEquipment[],
  slotConfig: Record<EquipmentSlot, number>
): { canInstall: boolean; reason?: string } {
  // Check class compatibility
  if (!equipment.compatibleClasses.includes(unit.unitClass)) {
    return { canInstall: false, reason: `Not compatible with ${unit.unitClass}` }
  }

  // Check slot availability
  const slotsUsed = currentEquipment.filter(e => e.slot === equipment.slot).length
  const slotsAvailable = slotConfig[equipment.slot] ?? 0
  if (slotsUsed >= slotsAvailable) {
    return { canInstall: false, reason: `No ${equipment.slot} slots available` }
  }

  // Check incompatible equipment
  if (equipment.requirements.incompatibleEquipment) {
    for (const incompatId of equipment.requirements.incompatibleEquipment) {
      if (currentEquipment.some(e => e.equipment.id.startsWith(incompatId))) {
        return { canInstall: false, reason: `Incompatible with ${incompatId}` }
      }
    }
  }

  // Check required equipment
  if (equipment.requirements.requiredEquipment) {
    for (const requiredId of equipment.requirements.requiredEquipment) {
      if (!currentEquipment.some(e => e.equipment.id.startsWith(requiredId))) {
        return { canInstall: false, reason: `Requires ${requiredId}` }
      }
    }
  }

  return { canInstall: true }
}

/**
 * Install equipment on unit
 */
export function installEquipment(
  equippedUnit: EquippedUnit,
  equipment: Equipment
): boolean {
  const check = canInstallEquipment(
    equipment,
    equippedUnit.unit,
    equippedUnit.equipment,
    equippedUnit.slotConfig
  )

  if (!check.canInstall) return false

  // Find next available slot index
  const slotsUsed = equippedUnit.equipment.filter(e => e.slot === equipment.slot)
  const slotIndex = slotsUsed.length

  equippedUnit.equipment.push({
    equipment,
    slot: equipment.slot,
    slotIndex,
    active: true,
  })

  equippedUnit.totalPowerConsumption += equipment.powerConsumption

  return true
}

/**
 * Uninstall equipment from unit
 */
export function uninstallEquipment(
  equippedUnit: EquippedUnit,
  equipmentId: string
): Equipment | null {
  const index = equippedUnit.equipment.findIndex(e => e.equipment.id === equipmentId)
  if (index === -1) return null

  const [removed] = equippedUnit.equipment.splice(index, 1)
  equippedUnit.totalPowerConsumption -= removed.equipment.powerConsumption

  return removed.equipment
}

/**
 * Calculate total stat modifiers from all equipment
 */
export function calculateEquipmentModifiers(
  equipment: InstalledEquipment[]
): EquipmentStatModifiers {
  const total: EquipmentStatModifiers = {
    damageMultiplier: 1,
    shieldMultiplier: 1,
    armorMultiplier: 1,
    hullMultiplier: 1,
    speedMultiplier: 1,
    accuracyBonus: 0,
    evasionBonus: 0,
    critChanceBonus: 0,
    critDamageBonus: 0,
    pointDefenseBonus: 0,
    shieldRegenBonus: 0,
    damageTypeBonuses: {},
    resistanceBonuses: {},
    crewCapacityBonus: 0,
    crewCombatBonus: 0,
    boardingPowerBonus: 0,
    antiBoardingBonus: 0,
  }

  for (const installed of equipment) {
    if (!installed.active) continue

    const mods = installed.equipment.statModifiers

    // Multiplicative bonuses
    if (mods.damageMultiplier) total.damageMultiplier! *= mods.damageMultiplier
    if (mods.shieldMultiplier) total.shieldMultiplier! *= mods.shieldMultiplier
    if (mods.armorMultiplier) total.armorMultiplier! *= mods.armorMultiplier
    if (mods.hullMultiplier) total.hullMultiplier! *= mods.hullMultiplier
    if (mods.speedMultiplier) total.speedMultiplier! *= mods.speedMultiplier

    // Additive bonuses
    if (mods.accuracyBonus) total.accuracyBonus! += mods.accuracyBonus
    if (mods.evasionBonus) total.evasionBonus! += mods.evasionBonus
    if (mods.critChanceBonus) total.critChanceBonus! += mods.critChanceBonus
    if (mods.critDamageBonus) total.critDamageBonus! += mods.critDamageBonus
    if (mods.pointDefenseBonus) total.pointDefenseBonus! += mods.pointDefenseBonus
    if (mods.shieldRegenBonus) total.shieldRegenBonus! += mods.shieldRegenBonus
    if (mods.crewCapacityBonus) total.crewCapacityBonus! += mods.crewCapacityBonus
    if (mods.crewCombatBonus) total.crewCombatBonus! += mods.crewCombatBonus
    if (mods.boardingPowerBonus) total.boardingPowerBonus! += mods.boardingPowerBonus
    if (mods.antiBoardingBonus) total.antiBoardingBonus! += mods.antiBoardingBonus

    // Damage type bonuses
    if (mods.damageTypeBonuses) {
      for (const [type, value] of Object.entries(mods.damageTypeBonuses)) {
        const key = type as keyof DamageTypes
        total.damageTypeBonuses![key] = (total.damageTypeBonuses![key] ?? 0) + (value ?? 0)
      }
    }

    // Resistance bonuses
    if (mods.resistanceBonuses) {
      for (const [type, value] of Object.entries(mods.resistanceBonuses)) {
        const key = type as keyof ResistanceTypes
        total.resistanceBonuses![key] = (total.resistanceBonuses![key] ?? 0) + (value ?? 0)
      }
    }
  }

  return total
}

/**
 * Apply equipment modifiers to a unit
 */
export function applyEquipmentToUnit(
  unit: AdvancedCombatUnit,
  equipment: InstalledEquipment[]
): void {
  const mods = calculateEquipmentModifiers(equipment)

  // Apply multiplicative modifiers
  unit.defense.shield.max = Math.floor(unit.defense.shield.max * (mods.shieldMultiplier ?? 1))
  unit.defense.shield.current = Math.floor(unit.defense.shield.current * (mods.shieldMultiplier ?? 1))
  unit.defense.armor.max = Math.floor(unit.defense.armor.max * (mods.armorMultiplier ?? 1))
  unit.defense.armor.current = Math.floor(unit.defense.armor.current * (mods.armorMultiplier ?? 1))
  unit.defense.hull.max = Math.floor(unit.defense.hull.max * (mods.hullMultiplier ?? 1))
  unit.defense.hull.current = Math.floor(unit.defense.hull.current * (mods.hullMultiplier ?? 1))

  // Apply damage multiplier
  unit.damage.ballistic = Math.floor(unit.damage.ballistic * (mods.damageMultiplier ?? 1))
  unit.damage.ionic = Math.floor(unit.damage.ionic * (mods.damageMultiplier ?? 1))
  unit.damage.explosive = Math.floor(unit.damage.explosive * (mods.damageMultiplier ?? 1))

  // Apply additive modifiers
  unit.stats.accuracy += mods.accuracyBonus ?? 0
  unit.stats.evasion += mods.evasionBonus ?? 0
  unit.stats.critChance += mods.critChanceBonus ?? 0
  unit.stats.critMultiplier += mods.critDamageBonus ?? 0
  unit.stats.pointDefense += mods.pointDefenseBonus ?? 0
  unit.defense.shield.regenRate += mods.shieldRegenBonus ?? 0

  // Apply damage type bonuses
  if (mods.damageTypeBonuses) {
    unit.damage.ballistic += mods.damageTypeBonuses.ballistic ?? 0
    unit.damage.ionic += mods.damageTypeBonuses.ionic ?? 0
    unit.damage.explosive += mods.damageTypeBonuses.explosive ?? 0
    unit.damage.hacking += mods.damageTypeBonuses.hacking ?? 0
    unit.damage.boarding += mods.damageTypeBonuses.boarding ?? 0
  }

  // Apply resistance bonuses
  if (mods.resistanceBonuses) {
    unit.resistances.ballistic_resistance += mods.resistanceBonuses.ballistic_resistance ?? 0
    unit.resistances.ionic_resistance += mods.resistanceBonuses.ionic_resistance ?? 0
    unit.resistances.explosive_resistance += mods.resistanceBonuses.explosive_resistance ?? 0
    unit.resistances.hack_defense += mods.resistanceBonuses.hack_defense ?? 0
    unit.resistances.anti_boarding += mods.resistanceBonuses.anti_boarding ?? 0
  }

  // Apply crew bonuses
  unit.crew.max += mods.crewCapacityBonus ?? 0
  unit.crew.combatStrength += mods.crewCombatBonus ?? 0
  unit.damage.boarding += mods.boardingPowerBonus ?? 0
  unit.resistances.anti_boarding += mods.antiBoardingBonus ?? 0
}

/**
 * Process equipment triggers for a specific event
 */
export function processEquipmentTriggers(
  equipment: InstalledEquipment[],
  triggerType: TriggerCondition['trigger'],
  context: {
    unit: AdvancedCombatUnit
    target?: AdvancedCombatUnit
    damage?: number
    random?: () => number
  }
): AbilityEffect[] {
  const random = context.random ?? Math.random
  const effects: AbilityEffect[] = []

  for (const installed of equipment) {
    if (!installed.active) continue

    for (const trigger of installed.equipment.triggers) {
      if (trigger.condition.trigger !== triggerType) continue

      // Check threshold conditions
      if (trigger.condition.threshold !== undefined) {
        if (triggerType === 'on_low_hull') {
          const hullPercent = (context.unit.defense.hull.current / context.unit.defense.hull.max) * 100
          if (hullPercent >= trigger.condition.threshold) continue
        }
      }

      // Check chance
      if (random() * 100 > trigger.condition.chance) continue

      effects.push(trigger.effect)
    }
  }

  return effects
}

/**
 * Use an active ability
 */
export function useAbility(
  equipment: Equipment,
  abilityId: string
): AbilityEffect | null {
  const ability = equipment.abilities.find(a => a.id === abilityId)
  if (!ability) return null
  if (ability.currentCooldown > 0) return null

  // Set cooldown
  ability.currentCooldown = ability.cooldown

  // Consume use for consumables
  if (equipment.activationType === 'consumable') {
    if (equipment.usesRemaining !== undefined) {
      equipment.usesRemaining--
    }
  }

  return ability.effect
}

/**
 * Tick ability cooldowns
 */
export function tickAbilityCooldowns(equipment: InstalledEquipment[]): void {
  for (const installed of equipment) {
    for (const ability of installed.equipment.abilities) {
      if (ability.currentCooldown > 0) {
        ability.currentCooldown--
      }
    }
  }
}

/**
 * Create equipped unit wrapper
 */
export function createEquippedUnit(
  unit: AdvancedCombatUnit,
  maxPower: number = 100
): EquippedUnit {
  return {
    unit,
    equipment: [],
    totalPowerConsumption: 0,
    maxPower,
    slotConfig: DEFAULT_SLOT_CONFIG[unit.unitClass] ?? DEFAULT_SLOT_CONFIG.utility,
  }
}

/**
 * Get all available abilities from equipped items
 */
export function getAvailableAbilities(
  equippedUnit: EquippedUnit
): Array<{ equipment: Equipment; ability: EquipmentAbility }> {
  const abilities: Array<{ equipment: Equipment; ability: EquipmentAbility }> = []

  for (const installed of equippedUnit.equipment) {
    if (!installed.active) continue

    for (const ability of installed.equipment.abilities) {
      if (ability.currentCooldown === 0) {
        // Check consumable uses
        if (
          installed.equipment.activationType === 'consumable' &&
          (installed.equipment.usesRemaining ?? 0) <= 0
        ) {
          continue
        }

        abilities.push({
          equipment: installed.equipment,
          ability,
        })
      }
    }
  }

  return abilities
}

File: unity_client/Assets/Scripts/SharedLib/Battle/EquipmentSystem.cs
// =============================================================================
// EquipmentSystem.cs — Ported from equipment-system.ts
// Ship equipment: slots, abilities, triggers, stat modifiers, functions
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace OGameX.SharedLib.Battle
{
    // =========================================================================
    // ENUMS
    // =========================================================================

    public enum EquipmentSlot
    {
        Weapon, Secondary, Shield, Armor, Engine, Computer, Special, Consumable
    }

    public enum EquipmentRarity
    {
        Common, Uncommon, Rare, Epic, Legendary
    }

    public enum ActivationType
    {
        Passive, Active, Triggered, Consumable
    }

    public enum AbilityEffectType
    {
        Damage, Buff, Debuff, Repair, Special
    }

    public enum TriggerType
    {
        OnHit, OnDamage, OnKill, OnCritical, OnShieldBreak,
        OnLowHull, OnRoundStart, OnRoundEnd, OnBoarding
    }

    // =========================================================================
    // STAT MODIFIERS
    // =========================================================================

    public class EquipmentStatModifiers
    {
        public double DamageMultiplier = 1;
        public double ShieldMultiplier = 1;
        public double ArmorMultiplier = 1;
        public double HullMultiplier = 1;
        public double SpeedMultiplier = 1;
        public double AccuracyBonus;
        public double EvasionBonus;
        public double CritChanceBonus;
        public double CritDamageBonus;
        public double PointDefenseBonus;
        public double ShieldRegenBonus;
        public DamageTypes DamageTypeBonuses;
        public ResistanceTypes ResistanceBonuses;
        public int CrewCapacityBonus;
        public double CrewCombatBonus;
        public double BoardingPowerBonus;
        public double AntiBoardingBonus;
    }

    // =========================================================================
    // ABILITY EFFECTS
    // =========================================================================

    public class AbilityEffect
    {
        public AbilityEffectType Type;
        public DamageTypes Damage;
        public int TargetCount;
        public bool AreaEffect;
        public double Penetration;
        public string Target;
        public EquipmentStatModifiers Modifiers;
        public int Duration;
        public StatusEffectType StatusEffect;
        public double Strength;
        public double RepairAmount;
        public string RepairType;
        public string EffectId;
        public Dictionary<string, object> Params;
    }

    public class EquipmentAbility
    {
        public string Id;
        public string Name;
        public string Description;
        public int Cooldown;
        public int CurrentCooldown;
        public int EnergyCost;
        public int Duration;
        public AbilityEffect Effect;
    }

    public class TriggerCondition
    {
        public TriggerType Trigger;
        public double? Threshold;
        public double Chance;
    }

    public class EquipmentTrigger
    {
        public TriggerCondition Condition;
        public AbilityEffect Effect;
    }

    public class EquipmentRequirements
    {
        public int WeaponsTech;
        public int ShieldTech;
        public int ArmorTech;
        public int ComputerTech;
        public int EngineTech;
        public int MinShipLevel;
        public List<string> RequiredEquipment;
        public List<string> IncompatibleEquipment;
    }

    // =========================================================================
    // EQUIPMENT
    // =========================================================================

    public class Equipment
    {
        public string Id;
        public string Name;
        public string Description;
        public EquipmentSlot Slot;
        public EquipmentRarity Rarity;
        public ActivationType ActivationType;
        public List<UnitClass> CompatibleClasses = new List<UnitClass>();
        public EquipmentStatModifiers StatModifiers = new EquipmentStatModifiers();
        public List<EquipmentAbility> Abilities = new List<EquipmentAbility>();
        public List<EquipmentTrigger> Triggers = new List<EquipmentTrigger>();
        public EquipmentRequirements Requirements = new EquipmentRequirements();
        public UnitCost InstallCost;
        public int PowerConsumption;
        public int Mass;
        public int? UsesRemaining;
    }

    public class InstalledEquipment
    {
        public Equipment Equipment;
        public EquipmentSlot Slot;
        public int SlotIndex;
        public bool Active = true;
    }

    public class EquippedUnit
    {
        public AdvancedCombatUnit Unit;
        public List<InstalledEquipment> Equipment = new List<InstalledEquipment>();
        public int TotalPowerConsumption;
        public int MaxPower;
        public Dictionary<EquipmentSlot, int> SlotConfig = new Dictionary<EquipmentSlot, int>();
    }

    // =========================================================================
    // SLOT CONFIG BY CLASS
    // =========================================================================

    public static class EquipmentSlotConfigs
    {
        public static Dictionary<EquipmentSlot, int> GetSlotConfig(UnitClass c)
        {
            return c switch
            {
                UnitClass.Fighter => S(1,0,1,1,1,0,0,1),
                UnitClass.Corvette => S(1,1,1,1,1,1,0,1),
                UnitClass.Frigate => S(2,1,1,1,1,1,1,2),
                UnitClass.Cruiser => S(2,2,2,2,1,1,1,2),
                UnitClass.Battlecruiser => S(3,2,2,2,1,2,1,2),
                UnitClass.Battleship => S(4,2,2,3,1,2,2,3),
                UnitClass.Dreadnought => S(5,3,3,3,1,2,3,3),
                UnitClass.Carrier => S(2,2,2,2,1,2,4,2),
                UnitClass.Transport => S(0,1,2,2,2,1,1,2),
                UnitClass.Platform => S(2,1,2,2,0,1,1,1),
                UnitClass.Turret => S(1,0,1,1,0,1,0,0),
                UnitClass.Defense => S(2,1,2,3,0,1,1,0),
                UnitClass.Missile => S(3,0,1,1,0,2,1,0),
                _ => S(0,1,1,1,2,2,1,0),
            };
        }

        private static Dictionary<EquipmentSlot, int> S(
            int w, int s2, int sh, int ar, int en, int co, int sp, int cn)
            => new Dictionary<EquipmentSlot, int>
            {
                [EquipmentSlot.Weapon]=w, [EquipmentSlot.Secondary]=s2,
                [EquipmentSlot.Shield]=sh, [EquipmentSlot.Armor]=ar,
                [EquipmentSlot.Engine]=en, [EquipmentSlot.Computer]=co,
                [EquipmentSlot.Special]=sp, [EquipmentSlot.Consumable]=cn,
            };
    }

    // =========================================================================
    // EQUIPMENT FUNCTIONS
    // =========================================================================

    public static class EquipmentFunctions
    {
        private static int _counter;

        public static (bool Ok, string Reason) CanInstall(
            Equipment eq, AdvancedCombatUnit unit,
            List<InstalledEquipment> current, Dictionary<EquipmentSlot, int> slots)
        {
            if (!eq.CompatibleClasses.Contains(unit.UnitClass))
                return (false, $"Not compatible with {unit.UnitClass}");
            int used = current.Count(e => e.Slot == eq.Slot);
            if (used >= slots.GetValueOrDefault(eq.Slot, 0))
                return (false, $"No {eq.Slot} slots available");
            if (eq.Requirements?.IncompatibleEquipment != null)
                foreach (var id in eq.Requirements.IncompatibleEquipment)
                    if (current.Any(e => e.Equipment.Id.StartsWith(id)))
                        return (false, $"Incompatible with {id}");
            if (eq.Requirements?.RequiredEquipment != null)
                foreach (var id in eq.Requirements.RequiredEquipment)
                    if (!current.Any(e => e.Equipment.Id.StartsWith(id)))
                        return (false, $"Requires {id}");
            return (true, null);
        }

        public static bool Install(EquippedUnit eu, Equipment eq)
        {
            var (ok, _) = CanInstall(eq, eu.Unit, eu.Equipment, eu.SlotConfig);
            if (!ok) return false;
            int idx = eu.Equipment.Count(e => e.Slot == eq.Slot);
            eu.Equipment.Add(new InstalledEquipment
                { Equipment = eq, Slot = eq.Slot, SlotIndex = idx, Active = true });
            eu.TotalPowerConsumption += eq.PowerConsumption;
            return true;
        }

        public static Equipment Uninstall(EquippedUnit eu, string equipmentId)
        {
            int i = eu.Equipment.FindIndex(e => e.Equipment.Id == equipmentId);
            if (i < 0) return null;
            var r = eu.Equipment[i];
            eu.Equipment.RemoveAt(i);
            eu.TotalPowerConsumption -= r.Equipment.PowerConsumption;
            return r.Equipment;
        }

        public static EquipmentStatModifiers CalculateModifiers(List<InstalledEquipment> equip)
        {
            var t = new EquipmentStatModifiers();
            foreach (var inst in equip)
            {
                if (!inst.Active) continue;
                var m = inst.Equipment.StatModifiers;
                t.DamageMultiplier *= m.DamageMultiplier;
                t.ShieldMultiplier *= m.ShieldMultiplier;
                t.ArmorMultiplier *= m.ArmorMultiplier;
                t.HullMultiplier *= m.HullMultiplier;
                t.SpeedMultiplier *= m.SpeedMultiplier;
                t.AccuracyBonus += m.AccuracyBonus;
                t.EvasionBonus += m.EvasionBonus;
                t.CritChanceBonus += m.CritChanceBonus;
                t.CritDamageBonus += m.CritDamageBonus;
                t.PointDefenseBonus += m.PointDefenseBonus;
                t.ShieldRegenBonus += m.ShieldRegenBonus;
                t.CrewCapacityBonus += m.CrewCapacityBonus;
                t.CrewCombatBonus += m.CrewCombatBonus;
                t.BoardingPowerBonus += m.BoardingPowerBonus;
                t.AntiBoardingBonus += m.AntiBoardingBonus;
                t.DamageTypeBonuses = t.DamageTypeBonuses + m.DamageTypeBonuses;
            }
            return t;
        }

        public static void ApplyToUnit(AdvancedCombatUnit u, List<InstalledEquipment> equip)
        {
            var m = CalculateModifiers(equip);
            u.Defense.Shield.Max = (int)(u.Defense.Shield.Max * m.ShieldMultiplier);
            u.Defense.Shield.Current = (int)(u.Defense.Shield.Current * m.ShieldMultiplier);
            u.Defense.Armor.Max = (int)(u.Defense.Armor.Max * m.ArmorMultiplier);
            u.Defense.Armor.Current = (int)(u.Defense.Armor.Current * m.ArmorMultiplier);
            u.Defense.Hull.Max = (int)(u.Defense.Hull.Max * m.HullMultiplier);
            u.Defense.Hull.Current = (int)(u.Defense.Hull.Current * m.HullMultiplier);
            u.Damage = new DamageTypes
            {
                Ballistic = (int)(u.Damage.Ballistic * m.DamageMultiplier) + (int)m.DamageTypeBonuses.Ballistic,
                Ionic = (int)(u.Damage.Ionic * m.DamageMultiplier) + (int)m.DamageTypeBonuses.Ionic,
                Explosive = (int)(u.Damage.Explosive * m.DamageMultiplier) + (int)m.DamageTypeBonuses.Explosive,
                Hacking = u.Damage.Hacking + (int)m.DamageTypeBonuses.Hacking,
                Boarding = u.Damage.Boarding + (int)(m.BoardingPowerBonus + m.DamageTypeBonuses.Boarding),
            };
            u.Stats.Accuracy += m.AccuracyBonus;
            u.Stats.Evasion += m.EvasionBonus;
            u.Stats.CritChance += m.CritChanceBonus;
            u.Stats.CritMultiplier += m.CritDamageBonus;
            u.Stats.PointDefense += m.PointDefenseBonus;
            u.Defense.Shield.RegenRate += (int)m.ShieldRegenBonus;
            u.Crew.Max += m.CrewCapacityBonus;
            u.Crew.CombatStrength += m.CrewCombatBonus;
            u.Resistances.AntiBoarding += m.AntiBoardingBonus;
        }

        public static List<AbilityEffect> ProcessTriggers(
            List<InstalledEquipment> equip, TriggerType trigger,
            AdvancedCombatUnit unit, Random rng = null)
        {
            rng ??= new Random();
            var fx = new List<AbilityEffect>();
            foreach (var inst in equip)
            {
                if (!inst.Active) continue;
                foreach (var tr in inst.Equipment.Triggers)
                {
                    if (tr.Condition.Trigger != trigger) continue;
                    if (tr.Condition.Threshold.HasValue && trigger == TriggerType.OnLowHull)
                    {
                        double hp = (double)unit.Defense.Hull.Current / unit.Defense.Hull.Max * 100;
                        if (hp >= tr.Condition.Threshold.Value) continue;
                    }
                    if (rng.NextDouble() * 100 > tr.Condition.Chance) continue;
                    fx.Add(tr.Effect);
                }
            }
            return fx;
        }

        public static AbilityEffect UseAbility(Equipment eq, string abilityId)
        {
            var ab = eq.Abilities.FirstOrDefault(a => a.Id == abilityId);
            if (ab == null || ab.CurrentCooldown > 0) return null;
            ab.CurrentCooldown = ab.Cooldown;
            if (eq.ActivationType == ActivationType.Consumable && eq.UsesRemaining.HasValue)
                eq.UsesRemaining--;
            return ab.Effect;
        }

        public static void TickCooldowns(List<InstalledEquipment> equip)
        {
            foreach (var inst in equip)
                foreach (var ab in inst.Equipment.Abilities)
                    if (ab.CurrentCooldown > 0) ab.CurrentCooldown--;
        }

        public static EquippedUnit CreateEquippedUnit(AdvancedCombatUnit unit, int maxPower = 100)
            => new EquippedUnit
            {
                Unit = unit, MaxPower = maxPower,
                SlotConfig = EquipmentSlotConfigs.GetSlotConfig(unit.UnitClass),
            };
    }
}
